_posts/2024-09-16-introducing-opensearch-2-17.md

```diff
@@ -0,0 +1,109 @@
+
+**Enhance query performance with efficient numeric term encoding and roaring bitmaps**
+
+The `terms` query type has been provided by OpenSearch since the beginning, serving as a method to filter searches by including documents that contain a field value within a specified set of values. It has always supported a _terms lookup_, which retrieves those filter values from another index based on a document ID, essentially allowing a limited version of cross-index join. A limitation has been that the number of values retrieved from the lookup index was limited to the hundreds (or potentially low thousands). In OpenSearch 2.17, we have [added support for encoding numeric term values](https://github.com/opensearch-project/OpenSearch/pull/14774) as a [roaring bitmap](https://roaringbitmap.org/). By encoding the values more efficiently, a search request can use a stored filter that matches over a million documents, with lower retrieval latency and less memory used. For details, see [Bitmap filtering](https://opensearch.org/docs/latest/query-dsl/term/terms/#bitmap-filtering).
+
+**Optimize query performance with improved multi-term query planning**
+
+In OpenSearch 2.12, we introduced a query planning feature for multi-term queries over keyword fields. These include `range` , `terms` , `wildcard` , `prefix`, and `fuzzy` queries. These queries tend be more expensive to evaluate on an index, because we need to collect all matching keyword terms for the field, then collect all documents that match all of the terms. With the change introduced in version 2.12, if one of these queries is part of a conjunction involving another clause that matches a few documents, we can avoid the heavy upfront work by applying the multi-term query as a secondary filter to documents that match the other clause. Unfortunately, [we have seen some cases](https://github.com/opensearch-project/OpenSearch/issues/14755) where this query planning has chosen the wrong path due to faulty cost estimation. While we work toward improving the cost estimates, we have reverted to the pre-2.12 behavior by default in version 2.17. You can enable the pre-2.12 behavior in versions 2.12--2.16 by setting `search.keyword_index_or_doc_values_enabled` to `true`.
```

Comment:
```suggestion
In OpenSearch 2.12, we introduced a query planning feature for multi-term queries over keyword fields. These include `range` , `terms` , `wildcard` , `prefix`, and `fuzzy` queries. These queries tend be more expensive to evaluate on an index because you first need to collect all matching keyword terms for the field and then collect all documents that match all of the terms. With the change introduced in version 2.12, if one of these queries is part of a conjunction involving another clause that matches a few documents, you can avoid the heavy upfront work by applying the multi-term query as a secondary filter to documents that match the other clause. Unfortunately, [we have seen some cases](https://github.com/opensearch-project/OpenSearch/issues/14755) in which this query planning has chosen the wrong path due to faulty cost estimation. While we work toward improving the cost estimates, we have reverted to the pre-2.12 behavior by default in version 2.17. You can enable the pre-2.12 behavior in versions 2.12--2.16 by setting `search.keyword_index_or_doc_values_enabled` to `true`.
```